### PR TITLE
Task casting cleanup

### PR DESCRIPTION
**Pull request by distantcam**
Added a task cast extension. Using this method is less overhead than using the async keyword.

### CODE REVIEW

**Comment by danielmarbach** [comment]
I think a more efficient and failure proof version would be something like

```
public static Task<TResult> Cast<TSource, TResult>(this Task<TSource> task) where TSource:TResult
        {
            var tcs = new TaskCompletionSource<TResult>();

            task.ContinueWith(t =>
                {
                    if (t.IsFaulted)
                    {
                        tcs.TrySetException(t.Exception.InnerExceptions);
                    }
                    else if (t.IsCanceled)
                    {
                        tcs.TrySetCanceled();
                    }
                    else
                    {
                        tcs.TrySetResult(t.Result);
                    }
                },
                TaskContinuationOptions.ExecuteSynchronously);

            return tcs.Task;
        }
```

**Comment by danielmarbach** [comment]
@Particular/nservicebus-maintainers we dropped the ball on this one. Can we come to a conclusion?

**Comment by timbussmann** [comment]
since this may improve the performance (very slightly?) I'm good with the changes. +1 on @danielmarbach improvement by `TaskContinuationOptions.ExecuteSynchronously`. No strong opinion on `TrySet` vs. `Set` though.

File: src/NServiceBus.Core/TaskEx.cs [hunk 1]
```diff
@@ -39,5 +39,19 @@ public static Task ThrowIfNull(this Task task)
 
             throw new Exception(TaskIsNullExceptionMessage);
         }
+
+        public static Task<TResult> Cast<TSource, TResult>(this Task<TSource> task) where TSource:TResult
+        {
+            var tcs = new TaskCompletionSource<TResult>();
+            task.ContinueWith(t => {
+                if (t.IsFaulted)
+                    tcs.SetException(t.Exception.InnerException);
```
Comment (timbussmann) [inline src/NServiceBus.Core/TaskEx.cs]: as in daniel's snippet, I think this should be `t.Exception.InnerExceptions`

**Comment by danielmarbach** [comment]
Assigned myself. Will fix it and see if it is worth to pull in

**Comment by danielmarbach** [comment]
```ini

BenchmarkDotNet=v0.9.6.0
OS=Microsoft Windows NT 6.2.9200.0
Processor=Intel(R) Core(TM) i7-4980HQ CPU @ 2.80GHz, ProcessorCount=8
Frequency=2728070 ticks, Resolution=366.5595 ns, Timer=TSC
HostCLR=MS.NET 4.0.30319.42000, Arch=32-bit RELEASE
JitModules=clrjit-v4.6.1055.0

Type=TaskCastWithAwaitVsCastExtension  Mode=Throughput  TargetCount=100  

```
            Method | Platform |    Median |    StdDev |  Gen 0 |  Gen 1 | Gen 2 | Bytes Allocated/Op |
------------------ |--------- |---------- |---------- |------- |------- |------ |------------------- |
 TaskCastWithAwait |      X64 | 2.5054 us | 0.0552 us | 523.77 | 104.75 |     - |             366.59 |
 TaskCastExtension |      X64 | 2.3242 us | 0.0352 us | 625.90 | 106.60 |  0.98 |             393.61 |
 TaskCastWithAwait |      X86 | 2.1732 us | 0.0753 us | 855.00 | 106.00 |     - |             190.31 |
 TaskCastExtension |      X86 | 1.9373 us | 0.2106 us | 961.00 | 107.00 |     - |             206.01 |


Code used

```
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Configs;
using BenchmarkDotNet.Diagnostics.Windows;
using BenchmarkDotNet.Exporters;
using BenchmarkDotNet.Jobs;
using BenchmarkDotNet.Running;

namespace TaskCastPerf
{
    class Program
    {
        static void Main(string[] args)
        {
            var summary = BenchmarkRunner.Run<TaskCastWithAwaitVsCastExtension>();
            Console.WriteLine(summary.ToString());
            Console.ReadLine();
        }
    }

    [Config(typeof(Config))]
    public class TaskCastWithAwaitVsCastExtension
    {

        private class Config : ManualConfig
        {
            public Config()
            {
                Add(MarkdownExporter.GitHub);
                Add(new MemoryDiagnoser());
                Add(new Job { Platform = Platform.X86 }.WithTargetCount(100));
                Add(new Job { Platform = Platform.X64 }.WithTargetCount(100));
            }
        }


        [Benchmark]
        public async Task TaskCastWithAwait()
        {
            await AwaitCastSimulator.Simulate().ConfigureAwait(false);
        }

        [Benchmark]
        public async Task TaskCastExtension()
        {
            await TaskCastSimulator.Simulate().ConfigureAwait(false);
        }
    }

    public static class Helper
    {
        static readonly List<int> SampleList = new List<int>();

        public static async Task<List<int>> YieldList()
        {
            await Task.Yield();
            return SampleList;
        }
    }

    public static class AwaitCastSimulator
    {

        public static async Task<IReadOnlyCollection<int>>  Simulate()
        {
            return await Helper.YieldList().ConfigureAwait(false);
        }
    }

    public static class TaskCastSimulator
    {
        public static Task<IReadOnlyCollection<int>> Simulate()
        {
            return Helper.YieldList().Cast<List<int>, IReadOnlyCollection <int>>();
        }
    }

    public static class TaskExtensions
    {
        public static Task<TResult> Cast<TSource, TResult>(this Task<TSource> task) where TSource : TResult
        {
            var tcs = new TaskCompletionSource<TResult>();

            task.ContinueWith(t =>
            {
                if (t.IsFaulted)
                {
                    tcs.TrySetException(t.Exception.InnerExceptions);
                }
                else if (t.IsCanceled)
                {
                    tcs.TrySetCanceled();
                }
                else
                {
                    tcs.TrySetResult(t.Result);
                }
            }, TaskContinuationOptions.ExecuteSynchronously);

            return tcs.Task;
        }
    }
}
```

**Comment by timbussmann** [comment]
based on @danielmarbach measurements I think the performance gain shown is not worth the price of the increased amount of allocations.

I suggest to close the PR and stay with the current solution. Any objections @distantcam ?

**Comment by distantcam** [comment]
Let me look into why it's doing more allocs first.

I should have explained where this came from. @SimonCropp had mentioned the AsyncFixer analyzers were throwing false warnings for these methods. I created this PR to get rid of those warnings.

**Comment by danielmarbach** [comment]
@distantcam cool do that. of course, my benchmark is stripped down and not representative for what we have in the pipeline.

**Comment by distantcam** [comment]
I'm not seeing the same amount of allocs that @danielmarbach got, but they are consistently higher than the async version.

```ini

BenchmarkDotNet=v0.9.6.0
OS=Microsoft Windows NT 6.2.9200.0
Processor=Intel(R) Core(TM) i7-3770 CPU @ 3.40GHz, ProcessorCount=8
Frequency=3330092 ticks, Resolution=300.2920 ns, Timer=TSC
HostCLR=MS.NET 4.0.30319.42000, Arch=32-bit RELEASE
JitModules=clrjit-v4.6.1080.0

Type=TaskCastWithAwaitVsCastExtension  Mode=Throughput  TargetCount=100

```

            Method | Platform |    Median |    StdDev |    Gen 0 |  Gen 1 | Gen 2 | Bytes Allocated/Op |
------------------ |--------- |---------- |---------- |--------- |------- |------ |------------------- |
 TaskCastWithAwait |      X64 | 1.7941 us | 0.0073 us | 2,384.31 | 107.52 |     - |             374.05 |
 TaskCastExtension |      X64 | 1.8067 us | 0.0128 us | 2,590.76 | 107.04 |     - |             401.02 |
 TaskCastWithAwait |      X86 | 1.6978 us | 0.0176 us | 1,282.00 |      - |     - |             189.40 |
 TaskCastExtension |      X86 | 1.5476 us | 0.0082 us | 1,394.75 |      - |     - |             208.18 |

My guess would be because the extension is using `TaskCompletionSource` which is a class, vs the async state machine which is a struct.

Anyway, I'm not attached to this PR as it was a syntax fix for an analyzer. Given the problem is only in 2 places in the code too we could probably just ignore the warning there and call it a day.

I'm ok with whatever the maintainers decide.

**Comment by bording** [comment]
Considering that @distantcam's results also show some mixed perf benefits on top of it increasing allocations, I'm :-1: for merging this.

**Comment by danielmarbach** [comment]
@distantcam thanks again for bringing this up. I'm closing this PR. Would it make sense to send in a PR which suppresses those two warnings?

**Pull request closed without merging**